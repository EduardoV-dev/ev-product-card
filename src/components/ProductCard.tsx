import React, { createContext, CSSProperties } from 'react';
import {
  IInitialValues,
  IOnChangeArgs,
  IProductCardHandlers,
  IProductContextProps,
  Product,
} from '../interfaces/interfaces';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

export interface IProps {
  // children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: IInitialValues;
  onChange?: (args: IOnChangeArgs) => void;
  product: Product;
  style?: CSSProperties;
  value?: number;
}

const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: IProps): JSX.Element => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div
        className={`${styles.productCard} ${className ?? ''}`}
        {...{ style }}
      >
        {children({
          counter,
          isMaxCountReached,

          maxCount: initialValues?.maxCount || Infinity,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
