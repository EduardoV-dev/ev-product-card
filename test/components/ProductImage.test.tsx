import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard, { ProductImage } from '../../src/components';
import { product1 } from '../fixtures';

describe('<ProductImage />', () => {
  it('Should render the component with the personalized image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <ProductImage img="https://via.placeholder.com/150" />}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('Should render the component with the product image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
