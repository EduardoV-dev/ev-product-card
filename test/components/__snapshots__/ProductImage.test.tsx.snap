// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> Should render the component with the personalized image 1`] = `
<div
  className="productCard "
>
  <img
    alt="Coffee Mug"
    className="productImg "
    src="https://via.placeholder.com/150"
  />
</div>
`;

exports[`<ProductImage /> Should render the component with the product image 1`] = `
<div
  className="productCard "
>
  <img
    alt="Coffee Mug"
    className="productImg "
    src=""
  />
</div>
`;
